azurerm/internal/services/datashare/resource_arm_data_share.go

```diff
@@ -0,0 +1,308 @@
+	accountId, err := parse.DataShareAccountID(d.Get("account_id").(string))
+	if err != nil {
+		return err
+	}
+
+	if d.IsNewResource() {
+		existing, err := client.Get(ctx, accountId.ResourceGroup, accountId.Name, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for present of existing DataShare %q (Resource Group %q / accountName %q): %+v", name, accountId.ResourceGroup, accountId.Name, err)
+			}
+		}
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_data_share", *existing.ID)
+		}
+	}
+
+	share := datashare.Share{
+		ShareProperties: &datashare.ShareProperties{
+			ShareKind:   datashare.ShareKind(d.Get("share_kind").(string)),
+			Description: utils.String(d.Get("description").(string)),
+			Terms:       utils.String(d.Get("terms").(string)),
+		},
+	}
+
+	if _, err := client.Create(ctx, accountId.ResourceGroup, accountId.Name, name, share); err != nil {
+		return fmt.Errorf("creating DataShare %q (Resource Group %q / accountName %q): %+v", name, accountId.ResourceGroup, accountId.Name, err)
+	}
+
+	resp, err := client.Get(ctx, accountId.ResourceGroup, accountId.Name, name)
+	if err != nil {
+		return fmt.Errorf("retrieving DataShare %q (Resource Group %q / accountName %q): %+v", name, accountId.ResourceGroup, accountId.Name, err)
+	}
+
+	if resp.ID == nil || *resp.ID == "" {
+		return fmt.Errorf("reading DataShare %q (Resource Group %q / accountName %q): ID is empty", name, accountId.ResourceGroup, accountId.Name)
+	}
+
+	d.SetId(*resp.ID)
+
+	if d.HasChange("snapshot_schedule") {
+		// only one dependent sync setting is allowed in one data share
+		if syncIterator, err := syncClient.ListByShareComplete(ctx, accountId.ResourceGroup, accountId.Name, name, ""); syncIterator.NotDone() {
+			if err != nil {
+				return fmt.Errorf("listing DataShare %q snapshot schedule (Resource Group %q / accountName %q): %+v", name, accountId.ResourceGroup, accountId.Name, err)
+			}
+			if syncName := syncIterator.Value().(datashare.ScheduledSynchronizationSetting).Name; syncName != nil && *syncName != "" {
+				syncFuture, err := syncClient.Delete(ctx, accountId.ResourceGroup, accountId.Name, name, *syncName)
+				if err != nil {
+					return fmt.Errorf("deleting DataShare %q snapshot schedule (Resource Group %q / accountName %q): %+v", name, accountId.ResourceGroup, accountId.Name, err)
+				}
+				if err = syncFuture.WaitForCompletionRef(ctx, syncClient.Client); err != nil {
+					return fmt.Errorf("waiting for DataShare %q snapshot schedule (Resource Group %q / accountName %q) to be deleted: %+v", name, accountId.ResourceGroup, accountId.Name, err)
+				}
+			}
+		}
+
+		if snapshotSchedule := expandAzureRmDataShareSnapshotSchedule(d.Get("snapshot_schedule").([]interface{})); snapshotSchedule != nil {
+			if _, err := syncClient.Create(ctx, accountId.ResourceGroup, accountId.Name, name, name, snapshotSchedule); err != nil {
```

Comment:
It seems like you can supply a name for the snapshot? presuming that it is a 1 datashare to 1 snapshot schedule (if it was 1 to many it should be a separate resource) we should add the name to the block to greatly simplify this logic & allow the user to specify the name.

Reply:
hi @katbyte , It's one-one relationship. I am wondering the meaning simplifying the logic. We still need to list complete and delete sync setting in createupdate, in case the use has changed the name, right? We still need to list complete and set` name` in read, right? Does it mean we only simplify in the delete method?

Reply:
you can use `d.GetChange` to get the old and new value